### PR TITLE
Tweaks for the BSM W boson pole mass prediction

### PR DESCRIPTION
**Pull request by Expander**

### CODE REVIEW

File: model_specific/SM/sm_mw.cpp [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+ * Calculates the prediction for the W boson mass in the Standard
+ * Model, using the fit formula Eq (45) from arXiv:1411.7040.
+ * 
+ * @param mh SM Higgs boson pole mass
+ * @param mt SM top quark pole mass
+ * @param as MS-bar alpha_s(MZ) in the SM with 5 quark flavours
+ * @param da5had hadronic contributison Delta alpha_{had}^{(5)}(MZ^2)
+ *
+ * The authors of arXiv:1411.7040 recommend used (Table 1):
+ *
```
Comment (pathron) [inline model_specific/SM/sm_mw.cpp:42]: What does this comment mean?  We are not recommending they use these values from 2014 are we?  Maybe 'recommend' is some pasted accident and should just be deleted?

Comment (Expander) [reply model_specific/SM/sm_mw.cpp:42]: With the comment I just wanted to give a hint on a possible input parameter set. I find that this sometimes help to get some first example working (and refine/update it later). Anyway, we are not recommending to use these values. These are just the values the authors of arXiv:1411.7040 used.

I can remove it, if you like.

**Comment by Expander** [comment]
There is one thing I'm not sure about: Currently I'm using the hard-coded value `Electroweak_constants::MH = 125.09 GeV` as input to the routine. Would it be better to use the SM-like pole mass predicted by the BSM model instead? (Which may be far away from the 125.09 GeV, though.)

File: model_specific/SM/sm_mw.cpp (hunk 1 above)
Comment (Expander) [reply model_specific/SM/sm_mw.cpp:43]: I've removed the statement in commit 08de09bd655b2ed39cd83739e2c34f7e7c8558a9

**Comment by Expander** [comment]
@wkotlarski : It seems the test `test/test_MSSM_FlexibleDecay.cpp` is broken, possibly because of the modified W boson mass. Since this test has hard-coded numbers, would it make sense to set the W mass in the test to the value before this PR?

**Comment by wkotlarski** [comment]
I'm not sure if I understand what you're proposing. The test in some sense has a hardcoded mW because we are reading in a hard coded SLHA spectrum. Your changes must have influenced some internal conversion formulas which makes all parameters to drift a little.

**Comment by Expander** [comment]
If I understand the test correctly, it reads the **input** parameters from the SLHA string and then runs
~~~.cpp
   MSSM_spectrum_generator<Two_scale> spectrum_generator;
   spectrum_generator.set_settings(settings);
   spectrum_generator.run(qedqcd, input);

   MSSM_slha m = std::get<0>(spectrum_generator.get_models_slha());
~~~
The execution of `.run(...)` uses the input parameters and performs the iteration between high and low scales. In this iteration the output parameters are indeed slightly modified due to my modification, that's right. However, mW is not fully fixed, only in terms of the DR-bar input parameters.

If I understand correctly you're saying that you want to use **all the SLHA output** in the decay calculation, right? I.e. you don't want to call `.run(...)`. You want to call 
~~~.cpp
MSSM_slha m(input); // input is already read from SLHA
slha_io.fill(m); // fill pole masses, DR-bar parameters into m
~~~
Could that be?

**Comment by wkotlarski** [comment]
This is more complicated. Doing just
```cpp
MSSM_slha m(input); // input is already read from SLHA
slha_io.fill(m);
```
gives
```
unknown location(0): fatal error: in "test_MSSM_FlexibleDecay": std::runtime_error: Error in H->ddbar: down quarks cannot be massless
```

**Comment by Expander** [comment]
I see. The SLHA only contains the pole masses, not the DR/MS-bar masses. I'll try to keep your `.run(...)` call, but afterwards overwrite the pole masses etc. from the SLHA string.

**Comment by Expander** [comment]
I have created PR #410 to change this: base the test on all parameters given in the SLHA string, i.e. don't call `.run(...)`.

**Comment by wkotlarski** [comment]
@Expander, could you please help me understand the origin of the entire problem? In files `test_*_FlexibleDecay.cpp` we have hardcoded strings containing a whole SLHA spectrum file. But it looks like they were actually not used (at least not fully). What was done was that we only read inputs from SLHA outputs and then run spectrum generator once more. So in the end this string should contain an LesHouches input, not an SLHA spectrum (is that correct?). Because you made changes to mW this shifted recalculated outputs. Your new solution reads everything from the SLHA output and only recalculates DR masses. This should mean that as long as formulas for running masses don't change, tests should be stable and independent of further changes to how numerical quantities are computed (if we ever change refernce mWSM formula again this shouldn't break tests, right?). So if I understood all of this correctly, than we can do 2 things:

1. What you did for the MSSM. Regenerate SLHA output with new mW and then proceed with the procedure above.
2. Don't regenerate the output. This would make them consistent with the old mW computation but not the new one and only adopt numerical values in tests.

Wouldn't it be better to do 2 for all models. In the future we might again update mWSM and if we stick to 2 than we wouldn't update the spectra again, they'd just be hardcoded forever.

**Comment by Expander** [comment]
@wkotlarski You are 100% correct and indeed we have these two options. In PR #410 my intention was to do option 2, i.e. take the current status of the `development` branch, re-generate the SLHA input+output and put that into the test as fixed string.

I agree with you option 2 this is indeed the best option.

**Comment by wkotlarski** [comment]
Jesus, I have no idea why did I do it like that. I propose to do the following: I'll keep the old spectra as they are, make changes as we decided (not calling spectrum generator etc) and only adopt numbers in tests and make PR to branch `feature-MW-tweaks`. #410 could then be dropped. Would that be ok for you?

**Comment by Expander** [comment]
Yes, this would be ok for me.
(Your proposed PR should in principle work for its own, I guess, right?)

**Comment by wkotlarski** [comment]
I think so. So it could be merged to `development` directly too. Would you prefer that?

**Comment by Expander** [comment]
> I think so. So it could be merged to `development` directly too. Would you prefer that?

If that is possible, I'd prefer this, as it is a bit cleaner (organizationally). :)

**Review by wkotlarski — APPROVED** [review]

**Review by wkotlarski — CHANGES_REQUESTED** [review]
This PR should only be merged **after** we make a 2.6.2 release.

**Comment by wkotlarski** [comment]
Should we merge the `development` into this branch to make the tests pass?

**Comment by Expander** [comment]
Yes, please!

**Comment by wkotlarski** [comment]
@Expander, really?
```
test failed: line 145: ss_average_time > fs_average_time: 0.126937 > 0.136811
```

**Comment by Expander** [comment]
:D 
There must be some logic error here:
~~~.cpp
TEST_GREATER(ss_average_time, fs_average_time);
~~~

**Comment by Expander** [comment]
Maybe we should disable this test, because the comparison with SOFTSUSY is no longer fair.

**Comment by wkotlarski** [comment]
```
test/test_SM_low_scale_constraint.cpp(137): error: in "test_low_scale_constraint": difference{0.0312866} between -m.get_mu2(){7564.5} and sm.get_mu2(){7335.0125869652666} exceeds 1e-14
```

**Comment by wkotlarski** [comment]
@Expander, I'm sorry but I think only you know how to fix the above error. I know you're busy so I'm writing only in case you've missed the above message.

**Comment by Expander** [comment]
Thanks for the hint, I've seen your remark. I'll try to find the time to fix this on Friday. Sorry for the delay.

**Comment by wkotlarski** [comment]
Thanks a lot!

**Comment by Expander** [comment]
Shall we merge this PR into development for our 2.7.0 release?

**Comment by pathron** [comment]
Yes we should :).

Let me quickly look through and submit review tomorrow.

**Review by pathron — APPROVED** [review]

**Pull request merged**